### PR TITLE
Part1

### PR DESCRIPTION
**Pull request by GlizzyProductions**

### CODE REVIEW

File: Dragon.h [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
 {
+    Dragon (std::string name_, int hp, int armor_, int attackDamage_ = 80 ) : 
+    Character( hp, armor_, attackDamage_ ),
+    name (name_)
+    {
+        
+    }
+    
+    const std::string& getName() override { return name; }
+    std::string getStats() override { return getCharacterStats(this); }
```
Comment (matkatmusic) [inline Dragon.h]: These should all be implemented in the cpp file, not the header file.
- [x] revise

File: DragonSlayer.cpp [hunk 2]
```diff
@@ -10,12 +10,16 @@ void DragonSlayer::attack(Character& other)
...
-        assert(false);
+
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so they should USE their attack item before attacking the dragon... 
         //note: they should only use the item if the dragon's hitpoints are > 0...
         //note: items are single-use only, so you need to reset it after use.  
         //look in the Character class for how the other item types are reset after use.
+
+        attackItem->use(this);
+        attackItem.reset();
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: What do we always do before using a pointer (`unique_ptr`)? 
also:
```
note: they should only use the item if the dragon's hitpoints are > 0...
```
- [x] revise

File: DragonSlayer.h [hunk 3]
```diff
@@ -1,8 +1,22 @@
...
+    DragonSlayer (std::string name_, int hp, int armor_, int attackDamage_ = 4 ) : 
+    Character( hp, armor_, attackDamage_ ),
+    name (name_)
+    {
+        helpfulItems = makeHelpfulItems(3);
+        defensiveItems = makeDefensiveItems(3);
+    }
+    
+    const std::string& getName() override { return name; }
+    std::string getStats() override { return getCharacterStats(this); }
```
Comment (matkatmusic) [inline DragonSlayer.h]: Move these implementations to the cpp file
- [x] revise

File: Dwarf.h [hunk 4]
```diff
@@ -1,7 +1,20 @@
...
-    const std::string name;
-};
\ No newline at end of file
+    Dwarf (std::string name_, int hp, int armor_, int attackDamage_ = 4 ) : 
+    Character( hp, armor_, attackDamage_ ),
+    name (name_)
+    {
+        helpfulItems = makeHelpfulItems(2);
+        defensiveItems = makeDefensiveItems(5);
+    }
```
Comment (matkatmusic) [inline Dwarf.h]: Move these implementations to the cpp file
- [x] revise

File: Item.h [hunk 5]
```diff
@@ -5,13 +5,17 @@
...
     virtual void use(Character*) = 0;
-    Item(const std::string& name_, int effect ) : name(name_), boost(effect) { }
+
+    Item( const std::string& name_, int effect );
     
     inline const std::string& getName() { return name; }
     inline int getBoost() const { return boost; }
+
 private:
+    char padding[8];
```
Comment (matkatmusic) [inline Item.h]: remove this.
- [x] revise

File: Paladin.h [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
 struct Paladin : Character
 {
+    Paladin (std::string name_, int hp, int armor_, int attackDamage_ = 10 ) : 
+    Character( hp, armor_, attackDamage_ ),
+    name (name_)
+    {
+        helpfulItems = makeHelpfulItems(2);
+        defensiveItems = makeDefensiveItems(6);
+        padding[0] = 0;
+    }
```
Comment (matkatmusic) [inline Paladin.h]: Move these implementations to the cpp file
- [x] revise

File: Utility.cpp [hunk 7]
```diff
@@ -43,66 +45,66 @@ std::string getCharacterStats(Character* ch)
...
-    else if( auto* ch = dynamic_cast<DragonSlayer*>(character))
+    else if( auto* ds = dynamic_cast<DragonSlayer*>(character))
     {
-        assert(false);
         //DragonSlayers get a 10x boost when attacking dragons, from their attack item.
         //so their attack item should boost their attack damage by a factor of 10
         //this means you need to GET the attack damage, multiply it by the item's boost, and BOOST the attackDamage with that multiplied value.  
         //check Character.h for available member functions you can use.
+      character->boostAttackDamage( item->getBoost() * character->getAttackDamage() );
     }
```
Comment (matkatmusic) [inline Utility.cpp]: Fix this indentation
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
You should not have any default argument values in the constructors you wrote.

File: DragonSlayer.cpp (hunk 2 above)
Comment (GlizzyProductions) [reply DragonSlayer.cpp]: always check if a pointer is nullptr first before using it!

File: Character.cpp [hunk 8]
```diff
@@ -1,117 +1,133 @@
...
+        c) the initial value of your stats is updated to reflect this boosted stat for the next time you defeat another
+    character.
+  */
+
+    boost(hitPoints, *initialHitPoints);
+    boost(armor, *initialArmorLevel);
+    boost(attackDamage, *initialAttackDamage);
+
+    std::cout << getName() << " defeated " << other.getName()
+              << " and leveled up!" << std::endl;
```
Comment (matkatmusic) [inline Character.cpp]: Fix this indentation
- [x] revise

File: Character.cpp [hunk 9]
```diff
@@ -1,117 +1,133 @@
...
-        
-        hitPoints -= damage;
-        if( hitPoints < 0 )
-        {
-            std::cout << getName() << " has died in battle!" << std::endl;
-            hitPoints = 0;
-        }
+        current = initial;
+        current *= 1.1;
+        initial = current;
```
Comment (matkatmusic) [inline Character.cpp]: This logic does not match the logic described in `attackInternal`
- [x] revise

File: Character.cpp [hunk 10]
```diff
@@ -1,117 +1,133 @@
...
+        if (hitPoints < 0) 
+        {
+            std::cout << getName() << " has died in battle!" << std::endl;
+            hitPoints = 0;
+        }
+    }
+    
+  printStats();
+  return hitPoints;
+}
```
Comment (matkatmusic) [inline Character.cpp]: This function did not need to be moved here.
This move has resulted in extra code needing to be reviewed that did not need reviewing.

Don't change code that does not need to be changed.

- [x] revert

File: DragonSlayer.cpp [hunk 11]
```diff
@@ -1,29 +1,60 @@
...
-        while( dragon->getHP() > 0 )
+        //look in the Character class for how the other item types are reset after use.*/
+            
+        if( attackItem != nullptr )
         {
+            if(dragon->getHP() > 0)
+            {
+            attackItem->use(this);
+            attackItem.reset();
+            }
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: Fix this indentation
- [x] revise

File: DragonSlayer.h [hunk 12]
```diff
@@ -1,8 +1,18 @@
...
+    std::string getStats() override;
     void attack(Character& other) override;
-private:
-    const std::string name;
-};
\ No newline at end of file
+
+    private:
+        const std::string name;
+        std::unique_ptr<AttackItem> attackItem { new AttackItem() };
```
Comment (matkatmusic) [inline DragonSlayer.h]: Fix this indentation
- [x] revise

File: Dwarf.h [hunk 13]
```diff
@@ -1,7 +1,15 @@
...
-    const std::string name;
-};
\ No newline at end of file
+    Dwarf (std::string name_, int hp, int armor_, int attackDamage_ = 4 );
+    
+    const std::string& getName() override;
+    std::string getStats() override;
+                 
+    private:
+        const std::string name;
```
Comment (matkatmusic) [inline Dwarf.h]: Fix this indentation
- [x] revise

File: Item.cpp [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+Item::Item(const std::string& name_, int effect) : name(name_), boost(effect)
+{
+
+}
+
+void Item::use(Character*)
+{
+
+}
```
Comment (matkatmusic) [inline Item.cpp]: Don't implement a base class pure virtual function.
They're pure virtual to force derived classes to HAVE to implement them.
You take that requirement away when you implement them in the base class.
- [x] revise

File: Paladin.h [hunk 15]
```diff
@@ -1,7 +1,15 @@
...
-    const std::string name;
-};
\ No newline at end of file
+    Paladin (std::string name_, int hp, int armor_, int attackDamage_ = 10 );
+    
+    const std::string& getName() override;
+    std::string getStats() override;
+
+    private:
+        const std::string name;
```
Comment (matkatmusic) [inline Paladin.h]: Fix this indentation
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: Character.cpp (hunk 10 above)
Comment (GlizzyProductions) [reply Character.cpp]: understood

File: DragonSlayer.h (hunk 12 above)
Comment (GlizzyProductions) [reply DragonSlayer.h]: fixed?

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Please fix the FIXMEs

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**